merkle-tree/concurrent/src/lib.rs

```diff
@@ -324,160 +323,108 @@ where
-                    current_node = H::hashv(&[&intersection_node, &current_node])?;
-                    self.rightmost_proof[intersection_index] = intersection_node;
-                }
-                Ordering::Greater => {
-                    current_node = compute_parent_node::<H>(
-                        &current_node,
-                        &self.rightmost_proof[i],
-                        node_index,
-                        i,
-                    )?;
-                }
-            }
+        if MAX_CHANGELOG == 0 {
+            return Err(ConcurrentMerkleTreeError::AppendOnly);
         }
 
-        changelog_entry.root = current_node;
-        changelog_entry.path = changelog_path;
-        changelog_entry.index = self.next_index;
+        let mut proof = proof.to_owned();
 
-        self.inc_current_changelog_index();
-        if let Some(changelog_element) = self
-            .changelog
-            .get_mut(self.current_changelog_index as usize)
-        {
-            *changelog_element = *changelog_entry;
-        }
-        self.inc_current_root_index();
-        *self
-            .roots
-            .get_mut(self.current_root_index as usize)
-            .ok_or(ConcurrentMerkleTreeError::RootsZero)? = current_node;
-
-        self.sequence_number = self.sequence_number.saturating_add(1);
-        self.next_index = self.next_index.saturating_add(1);
-        self.rightmost_leaf = *leaf;
-
-        Ok(())
+        self.update_proof(changelog_index, leaf_index, &mut proof)?;
+        self.validate_proof(old_leaf, leaf_index, &mut proof)?;
+        self.update_leaf_in_tree(new_leaf, leaf_index, &mut proof)
     }
 
     /// Appends a new leaf to the tree.
     pub fn append(
         &mut self,
         leaf: &[u8; 32],
     ) -> Result<ChangelogEntry<HEIGHT>, ConcurrentMerkleTreeError> {
-        let mut changelog_entry = ChangelogEntry::default();
-        self.append_with_changelog_entry(leaf, &mut changelog_entry)?;
+        let changelog_entries = self.append_batch(&[leaf])?;
+        let changelog_entry = changelog_entries
+            .first()
+            .ok_or(ConcurrentMerkleTreeError::EmptyChangelogEntries)?
+            .to_owned();
         Ok(changelog_entry)
     }
 
-    /// Appends a new batch of leaves to the tree.
+    /// Appends a batch of new leaves to the tree.
     pub fn append_batch(
         &mut self,
         leaves: &[&[u8; 32]],
-    ) -> Result<Vec<Box<ChangelogEntry<HEIGHT>>>, ConcurrentMerkleTreeError> {
-        let mut changelog_entries: Vec<Box<ChangelogEntry<HEIGHT>>> = Vec::new();
+    ) -> Result<Vec<ChangelogEntry<HEIGHT>>, ConcurrentMerkleTreeError> {
+        if (self.next_index as usize + leaves.len() - 1) >= 1 << HEIGHT {
+            return Err(ConcurrentMerkleTreeError::TreeFull);
+        }
+
+        let first_leaf_index = self.next_index;
+        // Buffer of Merkle paths.
+        let mut merkle_paths = MerklePaths::<H, HEIGHT>::new(leaves.len());
```

Comment:
what was the thought process behind switching to paths instead of doing change log entries directly?

I think it's good because this decouples the changelogs from the update process already which will make it easier to implement canopies I believe.

Reply:
I decided to keep the paths/changelog logic separate and self-contained, because I felt like keeping the loop (which fills previous changelog entries) in this function would make the code less readable.

And also, I feel like the `MerklePaths` becomes a data structure on its own.